core/src/test/java/com/graphhopper/storage/TurnCostStorageTest.java

```diff
@@ -52,12 +53,13 @@ public void setup() {
     // 2--3
     // |
     // 4
-    public static void initGraph(Graph g) {
-        g.edge(0, 1, 3, true);
-        g.edge(0, 2, 1, true);
-        g.edge(1, 3, 1, true);
-        g.edge(2, 3, 1, true);
-        g.edge(2, 4, 1, true);
+    public static void initGraph(GraphHopperStorage g) {
+        GHUtility.setProperties(Arrays.asList(
+                g.edge(0, 1).setDistance(3),
+                g.edge(0, 2).setDistance(1),
+                g.edge(1, 3).setDistance(1),
+                g.edge(2, 3).setDistance(1),
+                g.edge(2, 4).setDistance(1)), g.getEncodingManager(), true, true);
```

Comment:
I am not sure yet if this is an improvement compared to several calls to setProperties? Especially setting `speed=maxspeed/2`  seems not very obvious (or maybe I do not understand it yet)?

Reply:
The idea was basically to have an alternative to the previously used `graph.edge(from, to, dist, bothDir)` (like you said before it was pretty useful). So we set some default speed for *all* encoders. And `maxspeed/2` is approximatively the previously used default speed for all encoders.

We should rename the method as a first step. But I'm unsure if we should make it more explicit as it will then get more verbose?

Reply:
Can we not just remove it entirely?

Reply:
What do you want to remove entirely? The method? I would rather expand its usage and make speed+encoder more explicit? I like it more as the edge creation is less hidden within GHUtility calls.

Reply:
> What do you want to remove entirely? The method? 

Yes, `GHUtility.setProperties(List<EdgeIteratorState>...` and setting maxspeed/2

Reply:
... or maybe even (thinking out loud now):

```java
EdgeIteratorState edge = graph.edge(3, 5).setDistance(200).set(accessEnc, true, false).set(speedEnc, 60, 60);
```

i.e. add setter methods for encoded values to set both directions with one call.

Reply:
I think I favor the last option I proposed right now because it is very explicit but not even that verbose. I am just not sure how `set(encValue, boolean, boolean)` behaves for encoded values that are do not support two different directions (like road_class). Would be we be able to throw an error when we call `set(encValue, fwd, bwd)`, `encValue` is not bidirectional and `fwd!=bwd`?

Reply:
> `set(encValue, boolean, boolean)`

Sounds like a good solution.

> I am just not sure how set(encValue, boolean, boolean) behaves for encoded values that do not support two different directions

This should work the way you described

> Its only used in three places.

Only because I introduced it very late in the process as I stumbled over the fact that often multiple edges are initialized with the same speeds and IMO this "bulk" method improves readability of the edge creation.

What about a method like:

`setSpeeds(Double fwdSpeed, Double bwdSpeed, FlagEncoder encoder, EdgeIteratorState... edges)`

It is short and less verbose for many edges. 

And it would be already compatible to a version without "access"-bits. With that I mean the following: currently we have the "access"-bits defined only to use less bits (2 bits + speed bits intead of 2*speed_bits) and something that only the Weighting can define. Hmmh, or before or additionally to #1835 it might be interesting to investigate if we in general move away from access bits and just use `2*speed_bits`? So for higher precision of 7 bits this means 5 bits more per vehicle profile but even for hiking or walking this would make sense (#1679). Not sure.

Reply:
> setSpeeds(Double fwdSpeed, Double bwdSpeed, FlagEncoder encoder, EdgeIteratorState... edges)

Where would you put this method and can you make give an example for its usage? 

> And it would be already compatible to a version without "access"-bits. 

Ah because `fwd/bwdSpeed=infinity` would mean 'no access'? 

>  if we in general move away from access bits

This is a hard question, but so far it seems useful to me to e.g. separate motorized and outdoor profiles even without specifying the weighting already. Is this something we have to decide here? My guess is the refactoring here will not make it harder to remove the access flags. So yes we can introduce a helper method that already hides them, but we can just as well do this when we actually make this move? Delaying this until later seems to be the better option to me because we do not really know what we will do yet :)

Reply:
> Ah because fwd/bwdSpeed=infinity would mean 'no access'?

yes (for speed==0)

> Is this something we have to decide here?

No :) ... just thought that having a method without access bits specified is of an advantage for the future :)

> Delaying this until later seems to be the better option to me because we do not really know what we will do yet :)

ok

> Where would you put this method and can you give an example for its usage?

Hmmh, it does not work well with the encoder list it seems, but with one encoder it would be a bit more powerful than the existing setProperties method (as it has two speed parameters and allows multiple edges) and still is not more verbose. This method is used +1000 times and we can clean up some of its usages like in BridgeElevationInterpolatorTest, where we currently have:

```java
EdgeIteratorState edge01 = GHUtility.setProperties(graph.edge(0, 1).setDistance(10), encoder, 60, true, true);
EdgeIteratorState edge12 = GHUtility.setProperties(graph.edge(1, 2).setDistance(10), encoder, 60, true, true);
EdgeIteratorState edge23 = GHUtility.setProperties(graph.edge(2, 3).setDistance(10), encoder, 60, true, true);
EdgeIteratorState edge34 = GHUtility.setProperties(graph.edge(3, 4).setDistance(10), encoder, 60, true, true);
EdgeIteratorState edge56 = GHUtility.setProperties(graph.edge(5, 6).setDistance(10), encoder, 60, true, true);
EdgeIteratorState edge67 = GHUtility.setProperties(graph.edge(6, 7).setDistance(10), encoder, 60, true, true);
EdgeIteratorState edge78 = GHUtility.setProperties(graph.edge(7, 8).setDistance(10), encoder, 60, true, true);
EdgeIteratorState edge89 = GHUtility.setProperties(graph.edge(8, 9).setDistance(10), encoder, 60, true, true);
EdgeIteratorState edge17 = GHUtility.setProperties(graph.edge(1, 7).setDistance(10), encoder, 60, true, true);
EdgeIteratorState edge27 = GHUtility.setProperties(graph.edge(2, 7).setDistance(10), encoder, 60, true, true);
EdgeIteratorState edge37 = GHUtility.setProperties(graph.edge(3, 7).setDistance(10), encoder, 60, true, true);
```

and could replace it with:

```java
EdgeIteratorState edge01, edge12, edge23, edge34, edge56, edge67, edge78, edge89, edge17, edge27, edge37;
GHUtility.setProperties(60, 60, encoder,
        edge01 = graph.edge(0, 1).setDistance(10),
        edge12 = graph.edge(1, 2).setDistance(10),
        edge23 = graph.edge(2, 3).setDistance(10),
        edge34 = graph.edge(3, 4).setDistance(10),
        edge56 = graph.edge(5, 6).setDistance(10),
        edge67 = graph.edge(6, 7).setDistance(10),
        edge78 = graph.edge(7, 8).setDistance(10),
        edge89 = graph.edge(8, 9).setDistance(10),
        edge17 = graph.edge(1, 7).setDistance(10),
        edge27 = graph.edge(2, 7).setDistance(10),
        edge37 = graph.edge(3, 7).setDistance(10));
```

Reply:
Ok, either way seems fine. Is it a lot of work to turn the single edge creations into calls to `(GHUtility.setSpeeds(fwdSpeed, bwdSpeed, encoder, EdgeIteratorState...)`? I am happy with [this](https://github.com/graphhopper/graphhopper/pull/2202#discussion_r537024992) format and maybe only use `GHUtility.setSpeeds` later in places we feel it makes sense?

Reply:
Yes, it is a lot of work to migrate the current stuff. Probably we should follow a similar approach like migrating to junit5 :)

> I am happy with this format

Should I introduce the methods `set(encValue, fwd, bwd)` in this PR or later?

Reply:
> Yes, it is a lot of work to migrate the current stuff. Probably we should follow a similar approach like migrating to junit5 :)

I agree.

> Should I introduce the methods set(encValue, fwd, bwd) in this PR or later?

Is it harder or easier to search/replace GHUtility.setProperties here than it would be to search/replace graph.edge and do change it to set(encValue, fwd, bwd) in master? Hm yes why not do it here directly?

Reply:
When migrating the current code to:

```java
EdgeIteratorState edge = graph.edge(3, 5).setDistance(200).set(accessEnc, true, true).set(speedEnc, 60, 60);
```

there are two problems: if we throw an explicit error we cannot easily refactor the current usages as `speedEnc` is often not for two directions. Except we set the backward value in a silent manner which I did in https://github.com/graphhopper/graphhopper/commit/418454a9c9e070a283e1fab2037b2f11bd51f50f#diff-305b1d577b70e729c3845b447da93e1bc9668716a1c4b2e0928987b8623380b0R174, but I do not like it that much for our Java API.

And the code will be longer:

```java
EdgeIteratorState edge = graph.edge(3, 5).setDistance(200).set(encoder.getAccessEnc(), true, true).set(encoder.getSpeedEnc(), 60, 60);
```

So for now I would prefer to keep it (simpler?) like:

```java
EdgeIteratorState edge = GHUtilities.setSpeed(60, 60, encoder, graph.edge(3, 5).setDistance(200));
```

This code then uses the implicit setReverse:

```java
if (avgSpeedEnc.isStoreTwoDirections()) edge.setReverse(avgSpeedEnc, bwdSpeed);
```

which I find acceptable for the utility function and it has also the advantage that no access encoded value is mentioned and a speed of 0 would indicate that the direction is blocked.

Tricky stuff :)

Reply:
> Except we set the backward value in a silent manner 

Yeah that's not ideal, for example setting `.set(speedEnc, 60, 40)` should be an error if there is only one direction?
And I guess you right if we do just search&replace we a) do not know if the encoder is bidirectional and b) we do not even know if speed=60 is correct because that depends on the encoder? But maybe we can assume its a car encoder with one direction and speed 60 and only fix the remaining (hopefully not so many) cases?

> So for now I would prefer to keep it (simpler?) like: 

Yes sure if it makes the refactoring easier I would also use the utility method for now.

Reply:
> Yes sure if it makes the refactoring easier I would also use the utility method for now.

Yes, this 'regex refactoring' was relative time intense :)